sdk/servicebus/azure-servicebus/azure/servicebus/_pyamqp/client.py

```diff
@@ -895,6 +901,53 @@ def receive_message_batch(self, **kwargs):
         """
         return self._do_retryable_operation(self._receive_message_batch_impl, **kwargs)
 
+    def receive_messages_iter(self, on_message_received=None):
+        """Receive messages by generator. Messages returned in the generator have already been
+        accepted - if you wish to add logic to accept or reject messages based on custom
+        criteria, pass in a callback.
+
+        :param on_message_received: A callback to process messages as they arrive from the
+         service. It takes a single argument, a ~uamqp.message.Message object.
+        :type on_message_received: callable[~uamqp.message.Message]
+        """
+        self._message_received_callback = on_message_received
+        return self._message_generator()
+
+    def _message_generator(self):
+        """Iterate over processed messages in the receive queue.
+
+        :rtype: generator[~uamqp.message.Message]
+        """
+
+        auto_complete = self.auto_complete
+        self.auto_complete = False
+        receiving = True
+        message = None
+        try:
+
+            while receiving:
+
+                receiving = self.do_work()
+
+                message = self._received_messages.get(timeout=self._timeout)
```

Comment:
using queue timeout to close out iterator. When you hit the timeout you get a Empty exception which is caught by the except statement